src/Mutation.tsx

```diff
@@ -0,0 +1,237 @@
+import * as React from 'react';
+import * as PropTypes from 'prop-types';
+import ApolloClient, {
+  PureQueryOptions,
+  MutationUpdaterFn,
+  ApolloError,
+} from 'apollo-client';
+const invariant = require('invariant');
+import { DocumentNode } from 'graphql';
+const shallowEqual = require('fbjs/lib/shallowEqual');
+
+import { OperationVariables } from './types';
+import { parser, DocumentType } from './parser';
+
+export interface MutationResult<TData = any> {
+  data: TData;
+  error?: ApolloError;
+  loading: boolean;
+}
+
+export interface MutationProps<TData = any, TVariables = OperationVariables> {
+  mutation: DocumentNode;
+  variables?: TVariables;
+  optimisticResponse?: Object;
+  refetchQueries?: string[] | PureQueryOptions[];
+  update?: MutationUpdaterFn;
```

Comment:
I'm wondering if we should improve the type of the updater function so that it is parametric on `TData`, so that the second argument provides the result in the shape you'd expect, instead of an object of any keys of any value.
We could do this here in react apollo, the same way we did it to improve some of the `Query` types, and then we can send some PRs to correctly fix this in apollo client.

Reply:
I tried to improve on the type and made it parametric on TData, the second argument is now fully typed. The issue with this is that it is not so easy to to fix this issue. I had to override the type definition for `ExecutionResult` which is contained in graphql-js. So we will have to update the type in that repo to create a long-term sustainable solution I think,